[SPARK-19391][SparkR][ML] Tweedie GLM API for SparkR

actuaryzhang opened the pull request:
## What changes were proposed in this pull request?
Port Tweedie GLM  #16344  to SparkR

@felixcheung @yanboliang 

## How was this patch tested?
new test in SparkR

actuaryzhang committed:
start working on SparkR tweedie API

actuaryzhang committed:
set link only for non-tweedie; fix issue on aic

actuaryzhang committed:
add test for tweedie

SparkQA commented:
**[Test build #72111 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72111/testReport)** for PR 16729 at commit [`852dd6e`](https://github.com/apache/spark/commit/852dd6eec9ee960487f878dc332ea36acfd6102e).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72111/
Test FAILed.

actuaryzhang committed:
fix style

SparkQA commented:
**[Test build #72112 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72112/testReport)** for PR 16729 at commit [`5aa4ae7`](https://github.com/apache/spark/commit/5aa4ae79b62cd2039f8024f440c41f209a751090).

SparkQA commented:
**[Test build #72112 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72112/testReport)** for PR 16729 at commit [`5aa4ae7`](https://github.com/apache/spark/commit/5aa4ae79b62cd2039f8024f440c41f209a751090).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72112/
Test FAILed.

actuaryzhang committed:
fix style issue

SparkQA commented:
**[Test build #72114 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72114/testReport)** for PR 16729 at commit [`3682692`](https://github.com/apache/spark/commit/3682692752c1c332e29d2a92ca7fca23d8fbe2bc).

SparkQA commented:
**[Test build #72114 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72114/testReport)** for PR 16729 at commit [`3682692`](https://github.com/apache/spark/commit/3682692752c1c332e29d2a92ca7fca23d8fbe2bc).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72114/
Test FAILed.

felixcheung commented on R/pkg/inst/tests/testthat/test_mllib_regression.R:
```diff
@@ -77,6 +77,18 @@ test_that("spark.glm and predict", {
   expect_true(any(grepl("Dispersion parameter for gamma family", out)))
 
+  # tweedie family
+  require(statmod)
```
we can't require this as of now - we would need to update Jenkins otherwise it would fail like it is right now:
```
spark.glm and predict (@test_mllib_regression.R#81) - there is no package called 'statmod'

```

felixcheung replied:
in fact, `library` is more correct here as it will fail if the package isn't installed.

felixcheung commented on R/pkg/R/mllib_regression.R:
```diff
@@ -84,6 +84,12 @@ setClass("IsotonicRegressionModel", representation(jobj = "jobj"))
 #' summary(savedModel)
+#'
+#' # fit tweedie model
+#' require(statmod)
```
while it might be more correct, generally people use `library` instead of `require`

felixcheung commented:
I did look into this... I think it's great if `statmod` is there and we support it, but I'm concerned that we can't enable this `tweedie` family without an external dependency and the fact that we barely really depend on it. Is there a way to expose this in the API without having a hard dependency on tweedie family defined in `statmod`?

felixcheung commented on R/pkg/R/mllib_regression.R:
```diff
@@ -84,6 +84,12 @@ setClass("IsotonicRegressionModel", representation(jobj = "jobj"))
 #' # can also read back the saved model and print
 #' savedModel <- read.ml(path)
 #' summary(savedModel)
+#'
```
please update L56 for documentation. Also we should update the programming guide and vignettes too

felixcheung commented on R/pkg/R/mllib_regression.R:
```diff
@@ -109,7 +125,8 @@ setMethod("spark.glm", signature(data = "SparkDataFrame", formula = "formula"),
                                 "fit", formula, data@sdf, tolower(family$family), family$link,
-                                tol, as.integer(maxIter), as.character(weightCol), regParam)
+                                tol, as.integer(maxIter), as.character(weightCol), regParam,
+                                as.double(variancePower), as.double(linkPower))
```
we probably don't need to `as.double` here since it is either set to fixed values (L116) or from a calculation. Instead, we should check `var.power` and `link.power` are within the correct range - not sure if the tweedie function does that.

felixcheung commented on R/pkg/inst/tests/testthat/test_mllib_regression.R:
```diff
@@ -77,6 +77,18 @@ test_that("spark.glm and predict", {
+  prediction <- predict(model, training)
+  expect_equal(typeof(take(select(prediction, "prediction"), 1)$prediction), "double")
+  vals <- collect(select(prediction, "prediction"))
+  rVals <- suppressWarnings(predict(
```
why do we need `suppressWarnings` here?

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/r/GeneralizedLinearRegressionWrapper.scala:
```diff
@@ -81,15 +83,20 @@ private[r] object GeneralizedLinearRegressionWrapper
       .setFeaturesCol(rFormula.getFeaturesCol)
+    // set variancePower and linkPower if family is tweedie; otherwise, set link function
+    if (family.toLowerCase == "tweedie") {
+      glr = glr.setVariancePower(variancePower).setLinkPower(linkPower)
```
do we need to assign `glr =` here? generally the setter method will update the instance

felixcheung commented on R/pkg/inst/tests/testthat/test_mllib_regression.R:
```diff
@@ -77,6 +77,18 @@ test_that("spark.glm and predict", {
+  model <- spark.glm(training, Sepal_Width ~ Sepal_Length + Species,
+                     family = tweedie(var.power = 1.2, link.power = 1.0))
+  prediction <- predict(model, training)
+  expect_equal(typeof(take(select(prediction, "prediction"), 1)$prediction), "double")
```
you might want to use `dtypes` instead?

actuaryzhang replied:
Would you remind me what `dtypes` is and why we need to use it here? Thanks.

felixcheung commented on mllib/src/main/scala/org/apache/spark/ml/r/GeneralizedLinearRegressionWrapper.scala:
```diff
@@ -143,7 +150,12 @@ private[r] object GeneralizedLinearRegressionWrapper
     val rResidualDegreeOfFreedom: Long = summary.residualDegreeOfFreedom
-    val rAic: Double = summary.aic
+    val rAic: Double = if (family.toLowerCase == "tweedie" &&
+      !Array(0.0, 1.0, 2.0).contains(variancePower)) {
```
we are comparing double values here, do you know how reliable is this? should it have epsilon in the comparison?

actuaryzhang replied:
Thanks for the suggestion. Changed it to comparison instead.

actuaryzhang committed:
remove dependency on statmod

actuaryzhang committed:
create model matix directly from formula

actuaryzhang committed:
update glmWrapper

actuaryzhang committed:
add comments

SparkQA commented:
**[Test build #72131 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72131/testReport)** for PR 16729 at commit [`fb66ce0`](https://github.com/apache/spark/commit/fb66ce0bf2e4711a2679f0fef13eccdb27bb313c).

SparkQA commented:
**[Test build #72131 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72131/testReport)** for PR 16729 at commit [`fb66ce0`](https://github.com/apache/spark/commit/fb66ce0bf2e4711a2679f0fef13eccdb27bb313c).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72131/
Test FAILed.

actuaryzhang committed:
fix style issue

actuaryzhang commented:
@felixcheung Thanks so much for your quick and detailed review. I have made a new commit that removed dependency on `statmod` and fixed the issues you pointed out. The major change is to add `variancePower` and `linkPower` as arguments in `spark.glm` to avoid using the `tweedie` family from `statmod`. Let me know if this design is reasonable. Then I can update additional docs and the vignette. Thanks!

SparkQA commented:
**[Test build #72132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72132/testReport)** for PR 16729 at commit [`0d722fd`](https://github.com/apache/spark/commit/0d722fd6fe608fc6362aed0c0610b93e9afb7811).

SparkQA commented:
**[Test build #72132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72132/testReport)** for PR 16729 at commit [`0d722fd`](https://github.com/apache/spark/commit/0d722fd6fe608fc6362aed0c0610b93e9afb7811).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72132/
Test FAILed.

actuaryzhang committed:
remove statmod from suggest; update glm

SparkQA commented:
**[Test build #72136 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72136/testReport)** for PR 16729 at commit [`d11fc4b`](https://github.com/apache/spark/commit/d11fc4b90ffabb2c1cb2a72a9cb536430c8e2bd5).

actuaryzhang committed:
clean up doc

SparkQA commented:
**[Test build #72137 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72137/testReport)** for PR 16729 at commit [`4c24158`](https://github.com/apache/spark/commit/4c24158c442ab54f5acb777612cbadf22b5e7dcd).

SparkQA commented:
**[Test build #72136 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72136/testReport)** for PR 16729 at commit [`d11fc4b`](https://github.com/apache/spark/commit/d11fc4b90ffabb2c1cb2a72a9cb536430c8e2bd5).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72136/
Test FAILed.

SparkQA commented:
**[Test build #72137 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72137/testReport)** for PR 16729 at commit [`4c24158`](https://github.com/apache/spark/commit/4c24158c442ab54f5acb777612cbadf22b5e7dcd).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72137/
Test FAILed.

actuaryzhang committed:
remove link to statmod

SparkQA commented:
**[Test build #72144 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72144/testReport)** for PR 16729 at commit [`295711d`](https://github.com/apache/spark/commit/295711d313dd4e6a7a728a21261a5109b7e6e4a9).

SparkQA commented:
**[Test build #72144 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72144/testReport)** for PR 16729 at commit [`295711d`](https://github.com/apache/spark/commit/295711d313dd4e6a7a728a21261a5109b7e6e4a9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72144/
Test PASSed.

felixcheung commented on R/pkg/R/mllib_regression.R:
```diff
@@ -53,12 +53,19 @@ setClass("IsotonicRegressionModel", representation(jobj = "jobj"))
-#'               \code{Gamma}, and \code{poisson}.
+#'               \code{Gamma}, \code{poisson} and \code{"tweedie"}.
+#'               The tweedie family must be specified using character string \code{"tweedie"}. The family
+#'               function \code{tweedie} in the \code{statmod} package is not supported.
```
I prefer not having hard dependency on an external package in R, but if statmod is present, could we support `family = tweedie` (not string) too?

actuaryzhang committed:
allow R-like tweedie specification in family

actuaryzhang committed:
set default tweedie link to avoid passing functions to scala

actuaryzhang committed:
add tweedie in vignettes

actuaryzhang committed:
add internal tweedie family

SparkQA commented:
**[Test build #72259 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72259/testReport)** for PR 16729 at commit [`6737122`](https://github.com/apache/spark/commit/6737122763eef58c54a9b7a32994de6fe102c1d0).

SparkQA commented:
**[Test build #72259 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72259/testReport)** for PR 16729 at commit [`6737122`](https://github.com/apache/spark/commit/6737122763eef58c54a9b7a32994de6fe102c1d0).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72259/
Test FAILed.

actuaryzhang committed:
fix style

SparkQA commented:
**[Test build #72261 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72261/testReport)** for PR 16729 at commit [`0b5ed43`](https://github.com/apache/spark/commit/0b5ed434cc5ba68452f5ef556d20465fec4f8ccb).

actuaryzhang commented:
@felixcheung Great point! Yes, I think it's better to stick with the statmod syntax and allow the tweedie family to be specified as`tweedie(var.power, link.power)`. I tried a few ways to allow this and independence of statmod. The solution I came up with is creating and exporting a utility `tweedie` function within the SparkR package to mimic the statmod syntax. 

It will be better to avoid exporting it while still allowing users to use the syntax. We would need to use unevaluated expression, e.g., `substitute(family)`. This works in `spark.glm` by calling first `family = eval(substitute(family))` and then processing as before. However, it won't work with `glm`: `substitute` does not work in methods that have different arguments from the generic:  https://stat.ethz.ch/pipermail/r-help/2008-January/152296.html


I also updated the doc and added some examples and comments in the vignettes. Let me know what you think. Thanks much!

SparkQA commented:
**[Test build #72261 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72261/testReport)** for PR 16729 at commit [`0b5ed43`](https://github.com/apache/spark/commit/0b5ed434cc5ba68452f5ef556d20465fec4f8ccb).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72261/
Test FAILed.

actuaryzhang committed:
fix doc

SparkQA commented:
**[Test build #72268 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72268/testReport)** for PR 16729 at commit [`b10777e`](https://github.com/apache/spark/commit/b10777eb08621141df7190daa6157ff064c9d1af).

SparkQA commented:
**[Test build #72268 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72268/testReport)** for PR 16729 at commit [`b10777e`](https://github.com/apache/spark/commit/b10777eb08621141df7190daa6157ff064c9d1af).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72268/
Test FAILed.

actuaryzhang committed:
fix doc

actuaryzhang committed:
fix doc issue

SparkQA commented:
**[Test build #72273 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72273/testReport)** for PR 16729 at commit [`a9ac439`](https://github.com/apache/spark/commit/a9ac439d0e5d249f09cfe98d3aa25c75c22a820e).

SparkQA commented:
**[Test build #72273 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72273/testReport)** for PR 16729 at commit [`a9ac439`](https://github.com/apache/spark/commit/a9ac439d0e5d249f09cfe98d3aa25c75c22a820e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72273/
Test PASSed.

felixcheung commented:
ah thanks. so I were to 
```
library(statmod)
library(SparkR)
```

could I still access the statmod tweedie function?
ie. does statmod::tweedie still work with R base::glm?

actuaryzhang commented:
@felixcheung  The one from statmod will be masked and must be called using `statmod:tweedie`. 
We can copy the whole `tweedie` function from statmod into `SparkR` and this will avoid the issue (we will give credit to the original author). Does this sound reasonable?

felixcheung commented:
hmm.. I'm not sure we could do that because of Apache licensing policy (unless the owner decides to submit that here).

let me think about this a bit.

felixcheung commented:
@actuaryzhang thinking more about this I think compatibility would be fairly important, since `statmod` is fairly commonly used?
I hate to randomize you again, so how about we discuss our options a bit first?

- is there a way to not expose tweedie function outside the SparkR namespace? I mean if user have to explicit use it with `SparkR::tweedie` would be ok. My understanding is we don't, but it could be something I'm missing
- would it make sense to rename this, say `spark.tweedie`? Would that still be useable...
- ...or should we just give that up and add `var.power` `link.power` as parameter? I realize that is closer to what you have originally, but in addition to 2 new parameter we could also check if `tolower(family$family) == "tweedie"` and if so, get `family$var.power` instead, so in this case it would work when `statmod` is there, and it would still work when it is not. (but in addition, for consistency we would need to make passing `"tweedie"` as a string also work - but that wouldn't be much different from tweedie is a function)

What do you think?

actuaryzhang commented:
@felixcheung Thanks for the discussions. Will work on this in two weeks.

actuaryzhang committed:
merge pull

actuaryzhang committed:
make twedie internal (no export)

SparkQA commented:
**[Test build #73960 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73960/testReport)** for PR 16729 at commit [`6cbc62f`](https://github.com/apache/spark/commit/6cbc62fa8c8be37048193a6e3fc08195f1d497b5).

SparkQA commented:
**[Test build #73960 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73960/testReport)** for PR 16729 at commit [`6cbc62f`](https://github.com/apache/spark/commit/6cbc62fa8c8be37048193a6e3fc08195f1d497b5).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73960/
Test FAILed.

actuaryzhang committed:
fix test issue

SparkQA commented:
**[Test build #73964 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73964/testReport)** for PR 16729 at commit [`ef65adc`](https://github.com/apache/spark/commit/ef65adc69b710366266abf5df6f458af3a66a3bf).

SparkQA commented:
**[Test build #73964 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73964/testReport)** for PR 16729 at commit [`ef65adc`](https://github.com/apache/spark/commit/ef65adc69b710366266abf5df6f458af3a66a3bf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73964/
Test PASSed.

actuaryzhang commented:
@felixcheung Sorry for taking so long for this update. 

I think your first suggestion makes most sense, i.e., we do not expose the internal `tweedie`.
When `statmod` is loaded, users can use `tweedie` directly (from statmod); otherwise, they can use `SparkR::tweedie` which has the same syntax. 

I have made this to work. The following shows it now works both when statmod is not loaded (using `SparkR:::tweedie`) and when statmod is loaded (using `tweedie`).

Let me know if there is any other issues. Thanks. 

```
training <- suppressWarnings(createDataFrame(iris))
model1 <- spark.glm(training, Sepal_Width ~ Sepal_Length + Species,
                     family = SparkR:::tweedie(var.power = 1.2, link.power = 1.0))
summary(model1)$coefficients

                     Estimate Std. Error    t value     Pr(>|t|)
(Intercept)         1.7009666 0.22970461   7.405017 9.638512e-12
Sepal_Length        0.3436703 0.04518882   7.605206 3.200329e-12
Species_versicolor -0.9703190 0.07090188 -13.685377 0.000000e+00
Species_virginica  -0.9852650 0.09129919 -10.791607 0.000000e+00

library(statmod)
model2 <- spark.glm(training, Sepal_Width ~ Sepal_Length + Species,
                     family = tweedie(var.power = 1.2, link.power = 1.0))
summary(model2)$coefficients
                     Estimate Std. Error    t value     Pr(>|t|)
(Intercept)         1.7009666 0.22970461   7.405017 9.638512e-12
Sepal_Length        0.3436703 0.04518882   7.605206 3.200329e-12
Species_versicolor -0.9703190 0.07090188 -13.685377 0.000000e+00
Species_virginica  -0.9852650 0.09129919 -10.791607 0.000000e+00
```

felixcheung commented:
Thanks for working on this - to clarify, this only works with `SparkR:::tweedie` (ie. 3 `:`)
As this would be a private implementation?

actuaryzhang commented:
@felixcheung Yes, the SparkR `tweedie` is not exported. See below. 

```
 model1 <- spark.glm(training, Sepal_Width ~ Sepal_Length + Species,
+                     family = SparkR::tweedie(var.power = 1.2, link.power = 1.0))
Error: 'tweedie' is not an exported object from 'namespace:SparkR'
```

actuaryzhang commented:
@felixcheung Could you take a look at this new fix when you get a chance? Thanks.

felixcheung commented:
yea - I'm sorry if it was confusing - I was referring to `SparkR::tweedie` (2 `:`), and not `SparkR:::tweedie` (3 `:`), which was why I wasn't sure that could be done.

In the past there were concerns of exposing methods privately, so I'm not sure if we want to encourage accessing the tweedie function that way?

Perhaps then # 3 would be the only option (and it would be like Python)

actuaryzhang commented:
@felixcheung If we go with # 3, do we still want to compatibility with statmod::tweedie? It's  confusing to have two different ways of specifying the same model.

felixcheung commented:
@actuaryzhang that's true, it's not ideal.

This is somewhat an unusual case for R for several reasons.
In my head the guiding principles are:
- we avoid depending on another package (eg. `statmod`)
- we avoid naming a method/class that masks another *popular* package
- we can't ask user to call into private API
- we could use another package, for convenience, if it's loaded, but generally (*I'd agree*) we should avoid to minimize confusion or unpredictable behavior.

But since in this method we have this odd design where we take a R function (R `glm` family), and don't really use it, I thought it makes sense to allow the user to pass it in, that's why I suggested earlier as

> add var.power link.power as parameter? I realize that is closer to what you have originally, but in addition to 2 new parameters we could also check if tolower(family$family) == "tweedie" and if so, get family$var.power instead, so in this case it would work when statmod is there, and it would still work when it is not. (but in addition, for consistency we would need to make passing "tweedie" as a string also work - but that wouldn't be much different from tweedie is a function)

In the case `statmod` is loaded and the user is passing in `tweedie` in the family param, do we not need to explicitly check for it [here](https://github.com/apache/spark/pull/16729/files#diff-2e1e5d49cfda9702645a955f584572eeL96), since it is a valid function? IF we are going down this path, perhaps we should have whitelist instead of a blacklist?

Also the other concern is we do have another [signature `glm`](https://github.com/apache/spark/blob/master/R/pkg/R/mllib_regression.R#L147) where we "overload" the existing `stats::glm` one. In this case we can't add additional param like var.power or link.power, I think?

actuaryzhang committed:
add back variancePower and linkPower params

actuaryzhang committed:
update vignettes

actuaryzhang commented:
@felixcheung OK, new implementation of # 3. Now works in two ways:
1.  `family = "tweedie"` + `variancePower` + `linkPower`
2. When `statmod` is available, `tweedie()`

Please take another look. Thanks. 

```
# 1. Use variancePower and linkPower directly
> model <- spark.glm(training, Sepal_Width ~ Sepal_Length + Species,
+                      family = "tweedie", variancePower = 1.2, linkPower = 0.0)
> summary(model)$coefficients
                     Estimate Std. Error    t value     Pr(>|t|)
(Intercept)         0.6455411 0.07672839   8.413327 3.330669e-14
Sepal_Length        0.1169143 0.01508433   7.750714 1.425526e-12
Species_versicolor -0.3224752 0.02345653 -13.747781 0.000000e+00
Species_virginica  -0.3282173 0.03042303 -10.788450 0.000000e+00

# 2. Use statmod
> library(statmod)
> model <- spark.glm(training, Sepal_Width ~ Sepal_Length + Species, family = tweedie(1.2, 0))
> summary(model)$coefficients
                     Estimate Std. Error    t value     Pr(>|t|)
(Intercept)         0.6455411 0.07672839   8.413327 3.330669e-14
Sepal_Length        0.1169143 0.01508433   7.750714 1.425526e-12
Species_versicolor -0.3224752 0.02345653 -13.747781 0.000000e+00
Species_virginica  -0.3282173 0.03042303 -10.788450 0.000000e+00
```

SparkQA commented:
**[Test build #74242 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74242/testReport)** for PR 16729 at commit [`5ce4c84`](https://github.com/apache/spark/commit/5ce4c8414fabdaaa970931ac61fa53515d240503).

SparkQA commented:
**[Test build #74242 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74242/testReport)** for PR 16729 at commit [`5ce4c84`](https://github.com/apache/spark/commit/5ce4c8414fabdaaa970931ac61fa53515d240503).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74242/
Test FAILed.

actuaryzhang committed:
fix style

SparkQA commented:
**[Test build #74243 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74243/testReport)** for PR 16729 at commit [`aeeb3f7`](https://github.com/apache/spark/commit/aeeb3f780be846641a3f38eb127fcba3e3387cb8).

SparkQA commented:
**[Test build #74243 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74243/testReport)** for PR 16729 at commit [`aeeb3f7`](https://github.com/apache/spark/commit/aeeb3f780be846641a3f38eb127fcba3e3387cb8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74243/
Test PASSed.

actuaryzhang commented:
One other change I could make is to change `variancePower` and `linkPower` to `var.power` and `link.power` to be consistent with `statmod`. But l would like to get your feedback on this new design first.

felixcheung commented:
I like the example in this implementation! thanks
yes I think we should name them `var.power` and `link.power`.

felixcheung commented on R/pkg/R/mllib_regression.R:
```diff
@@ -100,6 +120,12 @@ setMethod("spark.glm", signature(data = "SparkDataFrame", formula = "formula"),
               stop("'family' not recognized")
             }
+            # Handle when family = statmod::tweedie()
+            if (tolower(family$family) == "tweedie" && !is.null(family$variance)) {
```
i assume it handles the "fake" family created on L111 correctly? it doesn't have `variance`

actuaryzhang replied:
This part only handles the case when `statmod::tweedie` is specified: it retrieves the `var.power` and `link.power` and construct a list with family name and link name to be used. 
The check for non-null `variance` is to skip handling the "fake" family. All we need when specifying `family = "tweedie"` is just a list with family name and link name.

felixcheung commented on R/pkg/R/mllib_regression.R:
```diff
@@ -84,14 +93,25 @@ setClass("IsotonicRegressionModel", representation(jobj = "jobj"))
-              family <- get(family, mode = "function", envir = parent.frame())
+              # Handle when family = "tweedie"
+              if (tolower(family) == "tweedie") {
+                family <- list(family = "tweedie", link = "linkNotUsed")
```
nit: I think you can set link = NULL

felixcheung commented on R/pkg/R/mllib_regression.R:
```diff
@@ -100,6 +120,12 @@ setMethod("spark.glm", signature(data = "SparkDataFrame", formula = "formula"),
+            if (tolower(family$family) == "tweedie" && !is.null(family$variance)) {
+              variancePower <- log(family$variance(exp(1)))
+              linkPower <- log(family$linkfun(exp(1)))
+              family <- list(family = "tweedie", link = "linkNotUsed")
```
ditto, link = NULL

felixcheung commented on R/pkg/R/mllib_regression.R:
```diff
@@ -84,14 +93,25 @@ setClass("IsotonicRegressionModel", representation(jobj = "jobj"))
+#'
+#' # fit tweedie model
+#' model <- spark.glm(df, Freq ~ Sex + Age, family = "tweedie",
+#'                    variancePower = 1.2, linkPower = 0)
```
could you add an example with statmod?

felixcheung reviewed (changes requested):
looking good, thanks for working on this!
you can follow up with programming guide change separately if you want.

actuaryzhang committed:
change names of tweedie parameters to be consistent with R

actuaryzhang commented:
@felixcheung Thanks for the feedback. Made a new commit that 

1. change `variancePower` and `linkPower` to `var.power` and `link.power`.
2. use `link = NULL` for tweedie family.
3. add example of using `statmod`.

Let me know if there is anything else needed.

SparkQA commented:
**[Test build #74417 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74417/testReport)** for PR 16729 at commit [`4cffc40`](https://github.com/apache/spark/commit/4cffc40f008641f59a17235ae67e4db0e975f03c).

SparkQA commented:
**[Test build #74417 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74417/testReport)** for PR 16729 at commit [`4cffc40`](https://github.com/apache/spark/commit/4cffc40f008641f59a17235ae67e4db0e975f03c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74417/
Test PASSed.

felixcheung reviewed (changes requested):
looking good! just this earlier comment https://github.com/apache/spark/pull/16729#discussion_r105566855

felixcheung commented on R/pkg/R/mllib_regression.R:
```diff
@@ -67,17 +54,20 @@ tweedie <- function(var.power = 0.0, link.power = 1.0 - var.power) {
+#'
+#'               Note that there are two ways to specify the tweedie family.
+#'               a) Set \code{family = "tweedie"} and specify the variancePower and linkPower
+#'               b) When package \code{statmod} is loaded, the tweedie family is specified using the
```
basically roxygen2 trims all the "insignificant whitespace"

actuaryzhang committed:
update doc

actuaryzhang commented:
Sorry that I forgot to address that comment. Fixed now.

SparkQA commented:
**[Test build #74423 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74423/testReport)** for PR 16729 at commit [`0b496a6`](https://github.com/apache/spark/commit/0b496a6a8880dfa81a0c8c3b00577b7313f93fbc).

SparkQA commented:
**[Test build #74423 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74423/testReport)** for PR 16729 at commit [`0b496a6`](https://github.com/apache/spark/commit/0b496a6a8880dfa81a0c8c3b00577b7313f93fbc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74423/
Test PASSed.

actuaryzhang commented:
@felixcheung Could you merge this please? Thanks!

felixcheung commented:
merged to master
